sqlalchemy_bigquery/base.py

```diff
@@ -831,6 +831,11 @@ def _process_time_partitioning(
         if time_partitioning.field is not None:
             field = time_partitioning.field
             if isinstance(
+                table.columns[time_partitioning.field].type,
+                sqlalchemy.sql.sqltypes.DATE,
+            ):
+                return f"PARTITION BY {field}"
+            elif isinstance(
                 table.columns[time_partitioning.field].type,
                 sqlalchemy.sql.sqltypes.TIMESTAMP,
```

Comment:
@tswast this PR is looking at adding logic to also handle partitioning on a date above and beyond using just DATE_TRUNC and TIMESTAMP_TRUNC functions.

[Looking at the breakdown](https://cloud.google.com/bigquery/docs/reference/standard-sql/timestamp_functions#timestamp_trunc) [of how](https://cloud.google.com/bigquery/docs/reference/standard-sql/date_functions#date_trunc) bigquery accepts time partitioning, it feels like we are missing a variety of the options/possibilities.

Am I missing something?

Reply:
@chalmerlowe can you clarify what you think is missing? it looks like the API expects one of
> Supported values are: * [HOUR](https://cloud.google.com/python/docs/reference/bigquery/latest/google.cloud.bigquery.table.TimePartitioningType) * [DAY](https://cloud.google.com/python/docs/reference/bigquery/latest/google.cloud.bigquery.table.TimePartitioningType) * [MONTH](https://cloud.google.com/python/docs/reference/bigquery/latest/google.cloud.bigquery.table.TimePartitioningType) * [YEAR](https://cloud.google.com/python/docs/reference/bigquery/latest/google.cloud.bigquery.table.TimePartitioningType)

so it seems like `TIMESTAMP_TRUNC` would work for all of those and `DATE` would work for `MONTH` and `YEAR`, so all that's missing is just plain `DATE`

Reply:
Column Data Type | HOUR | DAY | MONTH | YEAR
-- | -- | -- | -- | --
DATE | N/A | Fixed by https://github.com/googleapis/python-bigquery-sqlalchemy/pull/1057 | via DATE_TRUNC | via DATE_TRUNC
DATETIME | Currently via DATE_TRUNC TODO: use  DATETIME_TRUNC | Currently via DATE_TRUNC TODO: use  DATETIME_TRUNC | Currently via DATE_TRUNC TODO: use  DATETIME_TRUNC | Currently via DATE_TRUNC TODO: use  DATETIME_TRUNC
TIMESTAMP | via TIMESTAMP_TRUNC | via TIMESTAMP_TRUNC | via TIMESTAMP_TRUNC | via TIMESTAMP_TRUNC
_PARTITIONDATE | N/A | via DATE_TRUNC | via DATE_TRUNC | via DATE_TRUNC
_PARTITIONTIME | Not supported TODO: USE TIMESTAMP_TRUNC | Not supported TODO: USE TIMESTAMP_TRUNC | Not supported TODO: USE TIMESTAMP_TRUNC | Not supported TODO: USE TIMESTAMP_TRUNC

Reply:
@bnaul For clarity, my question was less about whether this PR is correct and more focused on the bigger question:

Since we have the function `_process_time_partitioning()` and it handles two use cases (soon to be three based on this PR), does it cover all the use cases that it ought to?

Tim and I looked at this. I had a matrix to help me ID what the function does/doesn't do/should do.

Tim cleaned it up and posted it above. There are capabilities that are satisfied via an existing BigQuery function (i.e. DATE_TRUNC). But there are use cases that are not fully satisfied OR not supported at all. Tim provides suggestions for how to resolve those gaps.

NOTE: **In no way am I saying you need to do all this work**. This is more for me as a maintainer to figure out what I want the end state to be either through this PR or a follow-on PR.

Reply:
ah yes that matrix is very helpful, thanks!